pkg/filesystem/subvolume.go

```diff
@@ -0,0 +1,139 @@
+limitations under the License.
+*/
+
+package subvolume
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	exec "github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type fsStruct struct {
+	Name string
+	data []string
+}
+
+type subVolumeInfo struct {
+	svg string
+	fs  string
+}
+
+func List(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace, includeStale string) map[string]subVolumeInfo {
+
+	if includeStale == "false" {
+		logging.Info("Listing subvolumes")
+	} else {
+		logging.Info("Listing stale subvolumes")
+	}
+	subvolumeNames := getK8sRefSubvolume(ctx, clientsets)
+	list := listCephFSSubvolumes(ctx, clientsets, operatorNamespace, clusterNamespace, includeStale == "true", subvolumeNames)
+	if len(list) == 0 && includeStale == "true" {
+		logging.Info("No stale subvolume found")
+	}
+	return list
+}
+
+// getk8sRefSubvolume returns the k8s ref for the subvolumes
+func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]subVolumeInfo {
+	pvList, err := clientsets.Kube.CoreV1().PersistentVolumes().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching PVs: %v\n", err))
+		os.Exit(1)
+	}
+	subvolumeNames := make(map[string]subVolumeInfo)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI.VolumeAttributes["subvolumeName"] != "" {
+			subvolumeNames[pv.Spec.CSI.VolumeAttributes["subvolumeName"]] = subVolumeInfo{}
+		}
+	}
+	return subvolumeNames
+}
+
+// listCephFSSubvolumes list all the subvolumes
+func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool, subvolumeNames map[string]subVolumeInfo) map[string]subVolumeInfo {
+
+	fsList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "ls", "--format", "json"}, operatorNamespace, clusterNamespace, true, false)
+	if fsList == "" {
+		logging.Fatal(fmt.Errorf("failed to get filesystem"))
+	}
+
+	fsstruct := unMarshaljson(fsList)
+	staleSubvolumeNames := make(map[string]subVolumeInfo)
+	var svList string
+
+	for _, fs := range fsstruct {
+		svgList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolumegroup", "ls", fs.Name, "--format", "json"}, operatorNamespace, clusterNamespace, true, false)
+		if len(svgList) == 0 {
+			logging.Fatal(fmt.Errorf("failed to get subvolumegroup for filesystem %q", fs.Name))
+		}
+		subvolg := unMarshaljson(svgList)
+		for _, svg := range subvolg {
+			svList = exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "ls", fs.Name, svg.Name}, operatorNamespace, clusterNamespace, true, false)
+			if len(svList) == 0 {
+				logging.Info("No subvolume found for filesystem %s and subvolumegroup %s", fs, svg)
+			}
+			subvol := unMarshaljson(svList)
```

Comment:
maybe you need to unmarshal before, and check `len(subvol)` (objects) instead of `len(svList)` (a string).